### PR TITLE
[V2 Streaming] Add generic Kernel Row <-> Spark Row zero-copy wrappers

### PR DESCRIPTION
**Pull request by seewishnew**
## 🥞 Stacked PR
Use this [link](https://github.com/delta-io/delta/pull/6701/files) to review incremental changes.
- [stack/streaming_phase0_pr1](https://github.com/delta-io/delta/pull/6700) [[Files changed](https://github.com/delta-io/delta/pull/6700/files)] [MERGED]
  - [**stack/streaming_phase0_pr2**](https://github.com/delta-io/delta/pull/6701) [[Files changed](https://github.com/delta-io/delta/pull/6701/files)]
    - [stack/streaming_phase0_pr3](https://github.com/delta-io/delta/pull/6702) [[Files changed](https://github.com/delta-io/delta/pull/6702/files/e3a7e2ae94fd8e7a6c342131faefc5d32b6240d5..6e056dcbcbf9d6868556d9ec628a1018794b849b)]
      - [stack/streaming_phase0_pr4](https://github.com/delta-io/delta/pull/6703) [[Files changed](https://github.com/delta-io/delta/pull/6703/files/6e056dcbcbf9d6868556d9ec628a1018794b849b..64188d2e77e0c9303a13670e0fbe7da7a029841b)]
        - [stack/streaming_phase0_pr5](https://github.com/delta-io/delta/pull/6706) [[Files changed](https://github.com/delta-io/delta/pull/6706/files/64188d2e77e0c9303a13670e0fbe7da7a029841b..9bf5431f800ae5cd6742ca25359931bf806819a8)]

---------
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

### Overall Scope (stacked PRs #6700-#6706)

This PR stack fixes a **driver OOM** in Delta V2 Structured Streaming when processing tables with very long commit histories during the initial snapshot. The existing implementation materializes the entire file listing on the driver, which exhausts memory when the commit history is large (100K+ files). The fix introduces a **distributed initial snapshot** approach that offloads the file listing and sorting to Spark executors via a DataFrame, keeping only metadata on the driver.

The stack is organized as:
- **PR1** (#6700, merged): `SerializableFileStatus` — make kernel `FileStatus` serializable for cross-executor shipping
- **PR2** (this PR): `KernelRowToSparkRow` / `SparkRowToKernelRow` — zero-copy row wrappers for Kernel<->Spark interop
- **PR3** (#6702): `SerializableReadOnlySnapshot`, `ScanFileRDD`, `DataFrameSnapshotCache` — the distributed data plane
- **PR4** (#6703): Integration into `SparkMicroBatchStream` and `SparkScan` behind a feature flag
- **PR5** (#6706): End-to-end integration tests

### This PR

Adds two generic utility classes for converting between Delta Kernel `Row` objects and Spark `Row` objects:

- **`KernelRowToSparkRow`**: Wraps a Kernel `Row` as a Spark `GenericRowWithSchema`, performing a deep copy of all field values at construction time. This ensures the Spark row remains valid after the underlying Kernel `ColumnarBatch` is closed. Handles all Kernel types including nested structs (recursive copy), arrays, maps, and date/timestamp conversions.

- **`SparkRowToKernelRow`**: Wraps a Spark `Row` as a Kernel `Row` (zero-copy, live view). Provides type-safe accessors with null guards (`Preconditions.checkState`) on all primitive getters to fail fast on null fields rather than throwing opaque NPEs. Handles `scala.collection.Map` vs `java.util.Map` polymorphism via `instanceof` checks.

Both classes are designed as general-purpose utilities reusable beyond the streaming use case.

## How was this patch tested?

- `KernelRowToSparkRowTest`: Tests round-trip conversion of all supported types (primitives, strings, binary, arrays, maps, nested structs), null handling, schema preservation, and deep-copy independence from the source row.
- `SparkRowToKernelRowTest`: Tests all typed getters, null guard behavior (throws on null primitive access), and correct handling of Spark-to-Kernel type mappings.

Run with: `build/sbt 'sparkV2/testOnly *KernelRowToSparkRowTest *SparkRowToKernelRowTest'`

## Does this PR introduce _any_ user-facing changes?

No. These are internal utility classes with no public API surface.

### CODE REVIEW

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+    } else if (dt instanceof ByteType) {
+      return accessor.getByte(ordinal);
+    } else if (dt instanceof ShortType) {
+      return accessor.getShort(ordinal);
+    } else if (dt instanceof IntegerType || dt instanceof DateType) {
+      return accessor.getInt(ordinal);
+    } else if (dt instanceof LongType
+        || dt instanceof TimestampType
+        || dt instanceof TimestampNTZType) {
+      return accessor.getLong(ordinal);
```
Comment (timothyw553) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java]: Spark's `Row.getDate` and `Row.getTimestamp` need `Date` and `Timestamp` objects, but we return `int` and `long` here, so they will crash, could we convert with `DateTimeUtils.toJavaDate`/`toJavaTimestamp`?

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/SparkRowToKernelRow.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    return new SparkRowToKernelRow(nested, nestedSchema);
+  }
+
+  @Override
+  public MapValue getMap(int ordinal) {
+    Map<?, ?> javaMap = sparkRow.getJavaMap(ordinal);
+    MapType mt = (MapType) kernelSchema.at(ordinal).getDataType();
+    return javaMapToKernelMapValue(javaMap, mt);
+  }
+
```
Comment (timothyw553) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/utils/SparkRowToKernelRow.java:218]: `sparkRow.getJavaMap` crashes if the value is a `java.util.Map` instead of a `scala.collection.Map`, could we check the type with `instanceof` first, like line 189?

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+  }
+
+  @Override
+  public Row copy() {
+    Object[] values = new Object[length()];
+    for (int i = 0; i < values.length; i++) {
+      values[i] = get(i);
+    }
+    return RowFactory.create(values);
+  }
```
Comment (timothyw553) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java:131]: nested structs still point at the live kernel row, so this copy fails once the `ColumnarBatch` closes, could we copy struct values recursively?

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/SparkRowToKernelRow.java [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+  @Override
+  public float getFloat(int ordinal) {
+    return sparkRow.getFloat(ordinal);
+  }
+
+  @Override
+  public double getDouble(int ordinal) {
+    return sparkRow.getDouble(ordinal);
+  }
```
Comment (timothyw553) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/utils/SparkRowToKernelRow.java:167]: these getters do not check `isNullAt` first, so a null Spark field throws `NullPointerException` from inside `sparkRow.getX`, could we add `if (isNullAt(ordinal))` at the start of each?

**Review by timothyw553 — COMMENTED** [review]
hey @seewishnew i just have a higher level question about the PR.. DSv2 reads use `InternalRow` directly, so Kernel -> Row -> InternalRow does the work twice (extra hop) and reintroduces the type-mapping issues called out below (with the Timestamp stuff), could we write `KernelRowToInternalRow` instead?

to my understanding after PR3, once Spark sort starts, Spark-side is already operating on `InternalRw` see https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/SortExec.scala

**Comment by seewishnew** [comment]
> i just have a higher level question about the PR.. DSv2 reads use InternalRow directly, so Kernel -> Row -> InternalRow does the work twice (extra hop) and reintroduces the type-mapping issues called out below (with the Timestamp stuff), could we write KernelRowToInternalRow instead?

But my understanding is that `InternalRow` is meant to be internal to Spark, with the Row being the public facing contract? 

Agreed there's this double-hop, but as an external lib I don't understand if it'd be okay for Delta to depend on Spark's internal representation of a row directly.

Also, in this specific case we'd be dealing with `AddFile.SCHEMA_WITHOUT_STATS` per (kernel/spark) row of the df. The per-row conversion cost overhead of these 6 fields should be super minimal:
```
  public static final StructType SCHEMA_WITHOUT_STATS =
      new StructType()
          .add("path", StringType.STRING, false /* nullable */)
          .add(
              "partitionValues",
              new MapType(StringType.STRING, StringType.STRING, true),
              false /* nullable*/)
          .add("size", LongType.LONG, false /* nullable*/)
          .add("modificationTime", LongType.LONG, false /* nullable*/)
          .add("dataChange", BooleanType.BOOLEAN, false /* nullable*/)
          .add("deletionVector", DeletionVectorDescriptor.READ_SCHEMA, true /* nullable */)
```

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/SparkRowToKernelRow.java (hunk 4 above)
Comment (seewishnew) [reply spark/v2/src/main/java/io/delta/spark/internal/v2/utils/SparkRowToKernelRow.java:167]: Good point, lemme add this 👍

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java (hunk 1 above)
Comment (seewishnew) [reply spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java]: Great catch! In general this is definitely worth expanding to treat these as true high-leverage util classes rather than a util class just for this streaming use case. Noted, will harden up these surface areas.

**Comment by timothyw553** [comment]
@seewishnew right, that is true. thanks for clarifying

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java (hunk 3 above)
Comment (seewishnew) [reply spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java:131]: Good point, updated to handle structs recursively

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/SparkRowToKernelRow.java [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+  public byte getByte(int ordinal) {
+    if (sparkRow.isNullAt(ordinal)) {
+      throw new IllegalStateException("Cannot read a null value as BYTE at ordinal " + ordinal);
+    }
+    return sparkRow.getByte(ordinal);
+  }
+
+  @Override
+  public short getShort(int ordinal) {
+    if (sparkRow.isNullAt(ordinal)) {
```
Comment (huan233usc) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/utils/SparkRowToKernelRow.java]: I think there should be a checkState method?

File: spark/v2/src/test/java/io/delta/spark/internal/v2/utils/SparkRowToKernelRowTest.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+import io.delta.kernel.data.MapValue;
+import io.delta.kernel.data.Row;
+import io.delta.kernel.internal.util.VectorUtils;
+import io.delta.kernel.types.*;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.spark.sql.RowFactory;
+import org.junit.jupiter.api.Test;
+
+public class SparkRowToKernelRowTest {
```
Comment (timothyw553) [inline spark/v2/src/test/java/io/delta/spark/internal/v2/utils/SparkRowToKernelRowTest.java:36]: could we add tests for date, timestamp, the null-throw guards, and `java.util.Map` input?

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+import java.util.Map;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.RowFactory;
+import org.apache.spark.sql.catalyst.util.DateTimeUtils;
+
+/**
+ * Adapts a Kernel Row to the Spark Row interface. Designed and tested for AddFile schema; other
+ * schemas may work but are not validated.
+ */
+public class KernelRowToSparkRow implements Row {
```
Comment (timothyw553) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java:36]: could we add an `.equals`/`.hashCode` test . im worried the scala trait defaults might lose to `Object.equals` since this is a java class?

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+    } else if (dt instanceof StringType) {
+      return accessor.getString(ordinal);
+    } else if (dt instanceof DecimalType) {
+      return accessor.getDecimal(ordinal);
+    } else if (dt instanceof BinaryType) {
+      return accessor.getBinary(ordinal);
+    } else if (dt instanceof StructType) {
+      return new KernelRowToSparkRow(
+          accessor.getStruct(ordinal),
+          SchemaUtils.convertKernelSchemaToSparkSchema((StructType) dt));
```
Comment (timothyw553) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java]: could we use `sparkSchema.apply(ordinal).dataType()` instead of re-converting the nested schema per row?

**Review by timothyw553 — COMMENTED** [review]
a few blockers from another pass.

**Review by timothyw553 — COMMENTED** [review]
almost there. a few comments.

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/SparkRowToKernelRow.java [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+    MapType mt = (MapType) kernelSchema.at(ordinal).getDataType();
+    return javaMapToKernelMapValue(javaMap, mt);
+  }
+
+  @Override
+  public ArrayValue getArray(int ordinal) {
+    if (sparkRow.isNullAt(ordinal)) {
+      return null;
+    }
+    List<?> javaList = sparkRow.getList(ordinal);
```
Comment (timothyw553) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/utils/SparkRowToKernelRow.java]: `sparkRow.getList(ordinal)` fails on a `java.util.List`. perhaps add the same `instanceof Seq` check used above?

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/SparkRowToKernelRow.java [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+        return DateTimeUtils.fromJavaTimestamp((java.sql.Timestamp) sparkValue);
+      } else if (sparkValue instanceof java.time.Instant) {
+        // Convert java.time.Instant -> epoch-microseconds long (UTC).
+        return DateTimeUtils.instantToMicros((java.time.Instant) sparkValue);
+      }
+      throw new UnsupportedOperationException(
+          "Cannot convert " + sparkValue.getClass() + " to TimestampType");
+    }
+    if (dt instanceof TimestampNTZType) {
+      if (sparkValue instanceof Long) {
```
Comment (timothyw553) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/utils/SparkRowToKernelRow.java]: this `Long` branch treats UTC micros and wall-clock micros as the same. maybe reject `Long` for `TimestampNTZType` and require `LocalDateTime`?

File: spark/v2/src/test/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRowTest.java [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+    ColumnVector ntzVector =
+        VectorUtils.buildColumnVector(List.of(ntzMicros), TimestampNTZType.TIMESTAMP_NTZ);
+
+    DefaultColumnarBatch batch =
+        new DefaultColumnarBatch(1, schema, new ColumnVector[] {dateVector, tsVector, ntzVector});
+    io.delta.kernel.data.Row kernelRow = batch.getRows().next();
+
+    Row sparkRow = new KernelRowToSparkRow(kernelRow);
+
+    Object dateVal = sparkRow.get(0);
```
Comment (timothyw553) [inline spark/v2/src/test/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRowTest.java:238]: test only checks the class, not the value, so a timezone bug in `toJavaDate` would pass. wdyt about `assertEquals` against the expected `Date` and `Timestamp`?

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+
+  public KernelRowToSparkRow(io.delta.kernel.data.Row kernelRow) {
+    this(kernelRow, SchemaUtils.convertKernelSchemaToSparkSchema(kernelRow.getSchema()));
+  }
+
+  /**
+   * Constructor that accepts a pre-computed Spark schema to avoid redundant schema conversion when
+   * wrapping many rows that share the same schema.
+   */
+  public KernelRowToSparkRow(
```
Comment (huan233usc) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java:51]: Why do we need this?

Comment (seewishnew) [reply spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java:51]: It's not strictly-speaking useful for this delta log OOM fix-- it was there in the original PoC and I presumed that it's a nice to have as a util method (as an analog to `SparkRowToKernelRow`).

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java [hunk 13]
```diff
@@ -0,0 +1,395 @@
+/*
+ * Copyright (2025) The Delta Lake Project Authors.
```
Comment (huan233usc) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java]: 26

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java [hunk 14]
```diff
@@ -0,0 +1,395 @@
...
+    return result;
+  }
+
+  @Override
+  public Row copy() {
+    Object[] values = new Object[length()];
+    for (int i = 0; i < values.length; i++) {
+      Object v = get(i);
+      // Recursively copy nested struct wrappers that still reference the live kernel row.
+      // Note: structs nested inside maps/arrays are already materialized by toSparkValue()
```
Comment (huan233usc) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java]: It is true for Map/Struct?

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java [hunk 15]
```diff
@@ -0,0 +1,395 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.RowFactory;
+import org.apache.spark.sql.catalyst.util.DateTimeUtils;
+
+/**
+ * Adapts a Kernel Row to the Spark Row interface. Designed and tested for AddFile schema; other
```
Comment (huan233usc) [inline spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java:33]: Do we consider only handle AddFile specifically here? If making it generic is lots of work and edge cases.

Comment (seewishnew) [reply spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java:33]: I think this comment is good to add here as a caveat. It's _designed_ and currently only used for the limited AddFile schema because that's all that matters for now, but its end-goal is to serve as something more generic and reusable; having this comment here helps inform future users that it's up to them to prove that this util works for their case with all the caveats for their domain.

**Review by timothyw553 — APPROVED** [review]

File: spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java (hunk 14 above)
Comment (seewishnew) [reply spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java]: (Forgot to submit reply 🤦‍♂️ ) Updated and added a test case to prove that it works as expected 👍

Comment (seewishnew) [reply spark/v2/src/main/java/io/delta/spark/internal/v2/utils/KernelRowToSparkRow.java]: Fyr: https://github.com/delta-io/delta/pull/6701/changes#diff-c99bdd424bcd87b31206dfd9b623b650fc50f89abbe5518e42c56175c50b539aR458-R476

**Review by huan233usc — APPROVED** [review]
